AppleSource/file_cmds-264.50.1/chmod/chmod.1
.\" Copyright (c) 1989, 1990, 1993, 1994
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" the Institute of Electrical and Electronics Engineers, Inc.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)chmod.1	8.4 (Berkeley) 3/31/94
.\" $FreeBSD: src/bin/chmod/chmod.1,v 1.33 2002/10/01 20:32:59 trhodes Exp $
.\"
.Dd July 08, 2004
.Dt CHMOD 1
.Os
.Sh NAME
.Nm chmod
.Nd change file modes or Access Control Lists
.Sh SYNOPSIS
.Nm chmod
.Op Fl fv
.Op Fl R Op Fl H | L | P
.Ar mode
.Ar
.Nm chmod
.Op Fl fv
.Op Fl R Op Fl H | L | P 
.Op -a | +a | =a
.Ar ACE
.Ar
.Nm chmod
.Op Fl fhv
.Op Fl R Op Fl H | L | P 
.Op Fl E
.Ar
.Nm chmod
.Op Fl fhv
.Op Fl R Op Fl H | L | P 
.Op Fl C
.Ar
.Nm chmod
.Op Fl fhv
.Op Fl R Op Fl H | L | P 
.Op Fl N
.Ar
.Sh DESCRIPTION
The
.Nm chmod
utility modifies the file mode bits of the listed files
as specified by the
.Ar mode
operand. It may also be used to modify the Access Control
Lists (ACLs) associated with the listed files.
.Pp
The generic options are as follows:
.Bl -tag -width Ds
.It Fl f
Do not display a diagnostic message if
.Nm chmod
could not modify the mode for
.Va file .
.It Fl H
If the
.Fl R
option is specified, symbolic links on the command line are followed.
(Symbolic links encountered in the tree traversal are not followed by
default.)
.It Fl h
If the file is a symbolic link, change the mode of the link itself
rather than the file that the link points to.
.It Fl L
If the
.Fl R
option is specified, all symbolic links are followed.
.It Fl P
If the
.Fl R
option is specified, no symbolic links are followed.
This is the default.
.It Fl R
Change the modes of the file hierarchies rooted in the files
instead of just the files themselves.
.It Fl v
Cause
.Nm chmod
to be verbose, showing filenames as the mode is modified.
If the
.Fl v
flag is specified more than once, the old and new modes of the file
will also be printed, in both octal and symbolic notation.
.El
.Pp
The
.Fl H ,
.Fl L
and
.Fl P
options are ignored unless the
.Fl R
option is specified.
In addition, these options override each other and the
command's actions are determined by the last one specified.
.Pp
Only the owner of a file or the super-user is permitted to change
the mode of a file.
.Sh DIAGNOSTICS
.Ex -std
.Sh MODES
Modes may be absolute or symbolic.
An absolute mode is an octal number constructed from the sum of
one or more of the following values:
.Pp
.Bl -tag -width 6n -compact -offset indent
.It Li 4000
(the set-user-ID-on-execution bit) Executable files with this bit set
will run with effective uid set to the uid of the file owner.
Directories with the set-user-id bit set will force all files and
sub-directories created in them to be owned by the directory owner
and not by the uid of the creating process, if the underlying file
system supports this feature: see
.Xr chmod 2
and the
.Ar suiddir
option to
.Xr mount 8 .
.It Li 2000
(the set-group-ID-on-execution bit) Executable files with this bit set
will run with effective gid set to the gid of the file owner.
.It Li 1000
(the sticky bit)
See
.Xr chmod 2
and
.Xr sticky 8 .
.It Li 0400
Allow read by owner.
.It Li 0200
Allow write by owner.
.It Li 0100
For files, allow execution by owner.
For directories, allow the owner to
search in the directory.
.It Li 0040
Allow read by group members.
.It Li 0020
Allow write by group members.
.It Li 0010
For files, allow execution by group members.
For directories, allow
group members to search in the directory.
.It Li 0004
Allow read by others.
.It Li 0002
Allow write by others.
.It Li 0001
For files, allow execution by others.
For directories allow others to
search in the directory.
.El
.Pp
For example, the absolute mode that permits read, write and execute by
the owner, read and execute by group members, read and execute by
others, and no set-uid or set-gid behaviour is 755
(400+200+100+040+010+004+001).
.Pp
The symbolic mode is described by the following grammar:
.Bd -literal -offset indent
mode         ::= clause [, clause ...]
clause       ::= [who ...] [action ...] action
action       ::= op [perm ...]
who          ::= a | u | g | o
op           ::= + | \- | =
perm         ::= r | s | t | w | x | X | u | g | o
.Ed
.Pp
The
.Ar who
symbols ``u'', ``g'', and ``o'' specify the user, group, and other parts
of the mode bits, respectively.
The
.Ar who
symbol ``a'' is equivalent to ``ugo''.
.Pp
The
.Ar perm
symbols represent the portions of the mode bits as follows:
.Pp
.Bl -tag -width Ds -compact -offset indent
.It r
The read bits.
.It s
The set-user-ID-on-execution and set-group-ID-on-execution bits.
.It t
The sticky bit.
.It w
The write bits.
.It x
The execute/search bits.
.It X
The execute/search bits if the file is a directory or any of the
execute/search bits are set in the original (unmodified) mode.
Operations with the
.Ar perm
symbol ``X'' are only meaningful in conjunction with the
.Ar op
symbol ``+'', and are ignored in all other cases.
.It u
The user permission bits in the original mode of the file.
.It g
The group permission bits in the original mode of the file.
.It o
The other permission bits in the original mode of the file.
.El
.Pp
The
.Ar op
symbols represent the operation performed, as follows:
.Bl -tag -width 4n
.It +
If no value is supplied for
.Ar perm ,
the ``+'' operation has no effect.
If no value is supplied for
.Ar who ,
each permission bit specified in
.Ar perm ,
for which the corresponding bit in the file mode creation mask
is clear, is set.
Otherwise, the mode bits represented by the specified
.Ar who
and
.Ar perm
values are set.
.It \&\-
If no value is supplied for
.Ar perm ,
the ``\-'' operation has no effect.
If no value is supplied for
.Ar who ,
each permission bit specified in
.Ar perm ,
for which the corresponding bit in the file mode creation mask
is clear, is cleared.
Otherwise, the mode bits represented by the specified
.Ar who
and
.Ar perm
values are cleared.
.It =
The mode bits specified by the
.Ar who
value are cleared, or, if no who value is specified, the owner, group
and other mode bits are cleared.
Then, if no value is supplied for
.Ar who ,
each permission bit specified in
.Ar perm ,
for which the corresponding bit in the file mode creation mask
is clear, is set.
Otherwise, the mode bits represented by the specified
.Ar who
and
.Ar perm
values are set.
.El
.Pp
Each
.Ar clause
specifies one or more operations to be performed on the mode
bits, and each operation is applied to the mode bits in the
order specified.
.Pp
Operations upon the other permissions only (specified by the symbol
``o'' by itself), in combination with the
.Ar perm
symbols ``s'' or ``t'', are ignored.
.Sh EXAMPLES OF VALID MODES
.Bl -tag -width "u=rwx,go=u-w" -compact
.It Li 644
make a file readable by anyone and writable by the owner only.
.Pp
.It Li go-w
deny write permission to group and others.
.Pp
.It Li =rw,+X
set the read and write permissions to the usual defaults, but
retain any execute permissions that are currently set.
.Pp
.It Li +X
make a directory or file searchable/executable by everyone if it is
already searchable/executable by anyone.
.Pp
.It Li 755
.It Li u=rwx,go=rx
.It Li u=rwx,go=u-w
make a file readable/executable by everyone and writable by the owner only.
.Pp
.It Li go=
clear all mode bits for group and others.
.Pp
.It Li g=u-w
set the group bits equal to the user bits, but clear the group write bit.
.El
.Sh ACL MANIPULATION OPTIONS
ACLs are manipulated using extensions to the symbolic mode
grammar.  Each file has one ACL, containing an ordered list of entries.
Each entry refers to a user or group, and grants or denies a set of
permissions.
In cases where a user and a group exist with the same name, the
user/group name can be prefixed with "user:" or "group:" in order to
specify the type of name.
.Pp
If the user or group name contains spaces you can use ':' as the delimiter
between name and permission.
.Pp
The following permissions are applicable to all filesystem objects:
.Bl -tag -width 6n -compact -offset indent
.It delete
Delete the item.  Deletion may be granted by either this permission
on an object or the delete_child right on the containing directory.
.It readattr
Read an objects basic attributes.  This is implicitly granted if 
the object can be looked up and not explicitly denied.
.It writeattr
Write an object's basic attributes.
.It readextattr
Read extended attributes.
.It writeextattr
Write extended attributes.
.It readsecurity
Read an object's extended security information (ACL).
.It writesecurity
Write an object's security information (ownership, mode, ACL).
.It chown
Change an object's ownership.
.El
.Pp
The following permissions are applicable to directories:
.Bl -tag -width 6n -compact -offset indent
.It list
List entries.
.It search
Look up files by name.
.It add_file
Add a file.
.It add_subdirectory
Add a subdirectory.
.It delete_child
Delete a contained object.  See the file delete permission above.
.El
.Pp
The following permissions are applicable to non-directory filesystem objects:
.Bl -tag -width 6n -compact -offset indent
.It read
Open for reading.
.It write
Open for writing.
.It append
Open for writing, but in a fashion that only allows writes into areas of 
the file not previously written.
.It execute
Execute the file as a script or program.
.El
.Pp
ACL inheritance is controlled with the following permissions words, which
may only be applied to directories:
.Bl -tag -width 6n -compact -offset indent
.It file_inherit
Inherit to files.
.It directory_inherit
Inherit to directories.
.It limit_inherit
This flag is only relevant to entries inherited by subdirectories; it
causes the directory_inherit flag to be cleared in the entry that is
inherited, preventing further nested subdirectories from also
inheriting the entry.
.It only_inherit
The entry is inherited by created items but not considered when processing
the ACL.
.El
.Pp
The ACL manipulation options are as follows:
.Bl -tag -width Ds
.It \fB+a\fR
The +a mode parses a new ACL entry from the next argument on
the commandline and inserts it into the canonical location in the
ACL. If the supplied entry refers to an identity already listed, the
two entries are combined.
.Pp
\fBExamples\fR
 # ls -le
 -rw-r--r--+ 1 juser  wheel  0 Apr 28 14:06 file1
 # chmod +a "admin allow write" file1
 # ls -le
 -rw-r--r--+ 1 juser  wheel  0 Apr 28 14:06 file1
   owner: juser
   1: admin allow write
 # chmod +a "guest deny read" file1
 # ls -le
 -rw-r--r--+ 1 juser  wheel  0 Apr 28 14:06 file1
   owner: juser
   1: guest deny read
   2: admin allow write
 # chmod +a "admin allow delete" file1
 # ls -le
 -rw-r--r--+ 1 juser  wheel  0 Apr 28 14:06 file1
   owner: juser
   1: guest deny read
   2: admin allow write,delete
 # chmod +a "User 1:allow:read" file
 # ls -le
 -rw-r--r--+ 1 juser  wheel  0 Apr 28 14:06 file1
   owner: juser
   1: guest deny read
   2: User 1 allow read
   3: admin allow write,delete
.Pp

CHANGELOG.md
# Version 5.028.1
	- Fixed Cmd+v shortcut
	- HostKey fingerprint as base64 encoding
	- Smoothing the HUD on resize or SplitView
	- Empty default user configuration bugs.
	- Zoom shortcuts changed.
	- Solarized theme fixes.
	
# Version 4.024.2
	- AirTerminals! Put a terminal on your remote AirPlay screen :)
	- Blink shortcuts. Use your external keyboard to move, create, or remove terminals. Configure the trigger too! View settings > Keyboard > Shortcuts for more info.
	- Want to use SmartKeys when an external keyboard is connected? Now you can switch them on and off from configuration.
	- View geometry of the screen from the overlay with each resize.
	- Switch cursor blinking on/off from Appearance settings.

	- Fixed TouchID issues when returning to the app.
	- Fixed cursor blink sequence so that it always starts as ON.
	- Fixed Host Port not resetting properly.
	- Fixed overlay getting covered in landscape mode.

	- Smoother experience thanks to cleanups and improvements.

# Version 3.021.2
	- iCloud Hosts sync. Synchronize hosts between devices. If a Host already has been synced, it provides conflict resolution. No critical data like passwords is saved.
	- Auto Lock. If enabled, when you lock/unlock your device, Blink will also be locked. Passcode and TouchID will be required to unlock the app.
	- Added ARMv7 support. Support for 32 bit devices like iPad 2, 3, iPhone 5, etc... We will publish depending on how well it performs!
	- Added IPv6 support for hosts.
	- Share Public Encryption Keys. You can now share the public key from the Keys section to other apps, like Mail.

	- Updated Fira Code font to v1.204.
	- Improved error checking on Themes and Font uploads. Auto correct if the GH URL is not a raw one.

	- Fixed bug with password not getting saved on host creation.
	- Fixed crash when hitting arrows with landscape keyboard on Plus devices.
	- Rolled back LC_CTYPE enforcement on server.

# Version 2.109 / 3.0
	- NEW Autorepeat for normal keys on external keyboard. Vim users rejoice!
	- NEW Default User for connections based on device.
	- NEW Comments for Public Keys.

	- Security updates to libraries.

	- Fixed critical bugs affecting ssh connections.

# Version 1.031
	- NEW On-screen keyboard with more space for modifiers, FKeys and Cursor keys. Redesigned for more space on the modifiers, and with a central scrollable area that handles more keys. Activate the Alternate keys by taping on the Alt key. And now tap on a modifier to activate it as a normal button, or make a long press to chain different combinations.
	- NEW Add your own Fonts & Themes! More info on (https://github.com/blinksh/fonts) and (https://github.com/blinksh/themes)
	- NEW Multistep authentication. Servers with google authenticator or similar will now connect without problems :)
	- NEW Fira Code font with ligatures included.

	- Added -l parameter to ssh for Hosts.
	- Improved message on how use hosts after adding them.
	- Fixed on-screen arrows.
	- Fixed F0 as F10 on external keyboard.
	- Fixed keys selection problem on settings
	- Fixed ssh-copy-id issue when accessing the host for the first time

# Version 1.019
	- Simplified build process.
	- New README and BUILD instructions

# Version 0.1020
	- "Get In Touch" and "About" sections completed.
	- iOS 10 support.
	- Added exception for iOS distribution.

	- Improved gesture support, specially on iPhone and iOS10.
	- MBProgressHUD updated with patch.
	- Test Terminal fixes for fonts and updates.
	- Disable font ligatures.
	- SSH Session termination fixes.

# Version 0.1010
	- Terminal customization! Customize your terminal with different default themes and fonts. Preview your changes within the Settings Preview section. Save your changes and restore them on each execution. Enjoy!

	- Create an id_rsa on first boot as default key.

	- Swipe and pinch conflicts
	- Copy changes to settings

# Version 0.931
	- Map Caps or Shift taps to Esc key.
	- Cursor keys (Home, End, PgUp, PgDown) with default to Cmd+arrow.
	- Function keys (F1-F10) with default to Cmd+number.
	- Change mapping of any function key to a different combination.

	- Improved placeholders for Hosts configurations.

	- Removed Alt as Esc from initial defaults.

# Version 0.927
	This version should complete the experience in relation to ssh and terminal configurations, mimicking a big part of what you can do in a normal shell:
	- Hosts Configuration. Preconfigure a host parameters, like user, port, key and commands.
	- Connect to a Host by specifying its name. Do "mosh plankton"
	- Overwrite parameters from host configuration from the shell. So carlos@plankton will override the user field on plankton.
	- Default Modifier keys settings changed: Ctrl is Ctrl, and Alt sends Esc. Everything else is undefined.
	- Modifier keys configuration. CAPS as Ctrl or ESC, no problem! Cmd as Ctrl? You have it! Configure everything to you liking.
	- Added secure passwords stored on Keychain to Host Configuration.
	- Exiting the session within the MCP closes the Space.

	- Fixed hang after "exec request accepted". Establishing connections should be smooth now.
	- Fixed adjustments on viewport after rotating the display.
	- Fixed Ctrl + Space sequences.
	- Fixed wrong/unexisting Ctrl sequences.
	- Added Esc+any character support.

	- New settings added to project. Will start to fill up in raw branch.

# Version 0.916
	This version contains many important bug fixes that should improve the experience a lot and reduce crashes to, hopefully, close to zero:
	- Unicode support to Blink shell. Thanks Yury!
	- Support for function keys + modifiers. Now you can do Ctrl/Alt/Shift + Arrows.
	- Improved experience with drag down to close. Let me know what you think :)
	- SSH with improved function keys in ncurses apps. htop, less, etc... should now work perfectly.
	- New help message with version and gestures.
	- Added asterisk, underscore and escape to SmartKeys.
	- Added HockeySDKResources.bundle to project for improved updates.


AppleSource/file_cmds-264.50.1/chmod/chmod.c
/*
 * Copyright (c) 1989, 1993, 1994
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__used static char const copyright[] =
"@(#) Copyright (c) 1989, 1993, 1994\n\r\
	The Regents of the University of California.  All rights reserved.\n\r";
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)chmod.c	8.8 (Berkeley) 4/1/94";
#endif
#endif /* not lint */
#include <sys/cdefs.h>
__RCSID("$FreeBSD: src/bin/chmod/chmod.c,v 1.27 2002/08/04 05:29:13 obrien Exp $");

#include <sys/types.h>
#include <sys/stat.h>

#include <err.h>
#include <errno.h>
#include <fts.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#ifdef __APPLE__
#include "chmod_acl.h"

#endif /*__APPLE__*/

int fflag = 0;

int chmod_main(int, char *[]);
void chmod_usage(void);

int
chmod_main(int argc, char *argv[])
{
    fflag = 0;
	FTS *ftsp = NULL;
	FTSENT *p = NULL;
	mode_t *set = NULL;
	long val = 0;
	int oct = 0;
	int Hflag, Lflag, Pflag, Rflag, ch, fts_options, hflag, rval;
	int vflag;
	char *ep, *mode;
	mode_t newmode, omode;
#ifdef __APPLE__
	unsigned int acloptflags = 0;
	long aclpos = -1;
	int inheritance_level = 0;
	int index = 0;
	size_t acloptlen = 0;
	int ace_arg_not_required = 0;
	acl_t acl_input = NULL;
#endif /* __APPLE__*/
	int (*change_mode)(const char *, mode_t);

	set = NULL;
	omode = 0;
	Hflag = Lflag = Pflag = Rflag = fflag = hflag = vflag = 0;
#ifndef __APPLE__
	while ((ch = getopt(argc, argv, "HLPRXfghorstuvwx")) != -1)
#else
	while ((ch = getopt(argc, argv, "ACEHILNPRVXafghinorstuvwx")) != -1)
#endif
		switch (ch) {
		case 'H':
			Hflag = 1;
			Lflag = 0;
			Pflag = 0;
			break;
		case 'L':
			Lflag = 1;
			Hflag = 0;
			Pflag = 0;
			break;
		case 'P':
			Hflag = Lflag = 0;
			Pflag = 1;
			break;
		case 'R':
			Rflag = 1;
			break;
		case 'f':
			fflag = 1;
			break;
		case 'h':
			/*
			 * In System V (and probably POSIX.2) the -h option
			 * causes chmod to change the mode of the symbolic
			 * link.  4.4BSD's symbolic links didn't have modes,
			 * so it was an undocumented noop.  In FreeBSD 3.0,
			 * lchmod(2) is introduced and this option does real
			 * work.
			 */
			hflag = 1;
			break;
#ifdef __APPLE__
		case 'a':
			if (argv[optind - 1][0] == '-' &&
			    argv[optind - 1][1] == ch)
				--optind;
			goto done;
		case 'A':
//			acloptflags |= ACL_FLAG | ACL_TO_STDOUT;
//			ace_arg_not_required = 1;
			warnx("-A not implemented");
                fprintf(stderr, "\r");
                return 0;
			goto done;
		case 'E':
			acloptflags |= ACL_FLAG | ACL_FROM_STDIN;
			goto done;
		case 'C':
			acloptflags |= ACL_FLAG | ACL_CHECK_CANONICITY;
			ace_arg_not_required = 1;
			goto done;
		case 'i':
			acloptflags |= ACL_FLAG | ACL_REMOVE_INHERIT_FLAG;
			ace_arg_not_required = 1;
			goto done;
		case 'I':
			acloptflags |= ACL_FLAG | ACL_REMOVE_INHERITED_ENTRIES;
			ace_arg_not_required = 1;
			goto done;
		case 'n':
			acloptflags |= ACL_FLAG | ACL_NO_TRANSLATE;
			break;
		case 'N':
			acloptflags |= ACL_FLAG | ACL_CLEAR_FLAG;
			ace_arg_not_required = 1;
			goto done;
		case 'V':
//			acloptflags |= ACL_FLAG | ACL_INVOKE_EDITOR;
//			ace_arg_not_required = 1;
			warn("-V not implemented");
                fprintf(stderr, "\r");
                return 0;
			goto done;
#endif /* __APPLE__ */
		/*
		 * XXX
		 * "-[rwx]" are valid mode commands.  If they are the entire
		 * argument, getopt has moved past them, so decrement optind.
		 * Regardless, we're done argument processing.
		 */
		case 'g': case 'o': case 'r': case 's':
		case 't': case 'u': case 'w': case 'X': case 'x':
			if (argv[optind - 1][0] == '-' &&
			    argv[optind - 1][1] == ch &&
			    argv[optind - 1][2] == '\0')
				--optind;
			goto done;
		case 'v':
			vflag++;
			break;
		case '?':
		default:
			chmod_usage();
            return 0;
		}
done:	argv += optind;
	argc -= optind;

#ifdef __APPLE__
    if (argc < ((acloptflags & ACL_FLAG) ? 1 : 2)) {
		chmod_usage();
        return 0;
    }
    if (!Rflag && (Hflag || Lflag || Pflag)) {
		warnx("options -H, -L, -P only useful with -R");
        fprintf(stderr, "\r");
    }
#else  /* !__APPLE__ */
    if (argc < 2) {
		chmod_usage();
        return 0; 
    }
#endif	/* __APPLE__ */

#ifdef __APPLE__
	if (!(acloptflags & ACL_FLAG) && ((acloptlen = strlen(argv[0])) > 1) && (argv[0][1] == 'a')) {
		acloptflags |= ACL_FLAG;
		switch (argv[0][0]) {
		case '+':
			acloptflags |= ACL_SET_FLAG;
			break;
		case '-':
			acloptflags |= ACL_DELETE_FLAG;
			break;
		case '=':
			acloptflags |= ACL_REWRITE_FLAG;
			break;
		default:
			acloptflags &= ~ACL_FLAG;
			goto apnoacl;
		}
		
        if (argc < 3) {
			chmod_usage();
            return 0;
        }

		if (acloptlen > 2) {
			for (index = 2; index < acloptlen; index++) {
				switch (argv[0][index]) {
				case '#':
					acloptflags |= ACL_ORDER_FLAG;

					if (argc < ((acloptflags & ACL_DELETE_FLAG)
                                ? 3 : 4)) {
						chmod_usage();
                        return 0;
                    }
					argv++;
					argc--;
					errno = 0;
					aclpos = strtol(argv[0], &ep, 0);

					if (aclpos > ACL_MAX_ENTRIES
					    || aclpos < 0)
						errno = ERANGE;
                        if (errno || *ep) {
						warn("Invalid ACL entry number: %ld", aclpos);
                            fprintf(stderr, "\r");
                            return 0;
                        }
					if (acloptflags & ACL_DELETE_FLAG)
						ace_arg_not_required = 1;

					goto apdone;
				case 'i':
					acloptflags |= ACL_INHERIT_FLAG;
					/* The +aii.. syntax to specify
					 * inheritance level is rather unwieldy,
					 * find an alternative.
					 */
                        inheritance_level++;
                        if (inheritance_level > 1) {
                            warnx("Inheritance across more than one generation is not currently supported");
                            fprintf(stderr, "\r");
                        }
					if (inheritance_level >= MAX_INHERITANCE_LEVEL)
						goto apdone;
					break;
				default:
					errno = EINVAL;
					chmod_usage();
                    return 0;
				}
			}
		}
apdone:
		argv++;
		argc--;
	}
apnoacl:
#endif /*__APPLE__*/

	if (Rflag) {
		fts_options = FTS_PHYSICAL;
        if (hflag) {
			warn("the -R and -h options may not be specified together.");
            fprintf(stderr, "\r");
            return 0;
        }
		if (Hflag)
			fts_options |= FTS_COMFOLLOW;
		if (Lflag) {
			fts_options &= ~FTS_PHYSICAL;
			fts_options |= FTS_LOGICAL;
		}
	} else
		fts_options = hflag ? FTS_PHYSICAL : FTS_LOGICAL;

	if (hflag)
		change_mode = lchmod;
	else
		change_mode = chmod;
#ifdef __APPLE__
	if (acloptflags & ACL_FROM_STDIN) {
		ssize_t readval = 0;
		size_t readtotal = 0;
		
		mode = (char *) malloc(MAX_ACL_TEXT_SIZE);
		
        if (mode == NULL) {
			warn("Unable to allocate mode string");
            fprintf(stderr, "\r");
            return 0;
        }
		/* Read the ACEs from STDIN */
		do {
			readtotal += readval;
			readval = read(STDIN_FILENO, mode + readtotal, 
				       MAX_ACL_TEXT_SIZE);
		} while ((readval > 0) && (readtotal <= MAX_ACL_TEXT_SIZE));
			
        if (0 == readtotal) {
			warn("-E specified, but read from STDIN failed");
            fprintf(stderr, "\r");
            return 0;
        }
		else
			mode[readtotal - 1] = '\0';
		--argv;
	}
	else
#endif /* __APPLE */
		mode = *argv;

#ifdef __APPLE__
	if ((acloptflags & ACL_FLAG)) {

		/* Are we deleting by entry number, verifying
		 * canonicity or performing some other operation that
		 * does not require an input entry? If so, there's no
		 * entry to convert.
		 */
		if (ace_arg_not_required) {
			--argv;
		}
		else {
                        /* Parse the text into an ACL*/
			acl_input = parse_acl_entries(mode);
			if (acl_input == NULL) {
				warn("Invalid ACL specification: %s", mode);
                fprintf(stderr, "\r");
                return 0;
            }
		}
	}
	else {
#endif /* __APPLE__*/
		if (*mode >= '0' && *mode <= '7') {
			errno = 0;
			val = strtol(mode, &ep, 8);
			if (val > USHRT_MAX || val < 0)
				errno = ERANGE;
            if (errno) {
				warn("Invalid file mode: %s", mode);
                fprintf(stderr, "\r");
                return 0;
            }
            if (*ep) {
				warn("Invalid file mode: %s", mode);
                fprintf(stderr, "\r");
                return 0;
            }
			omode = (mode_t)val;
			oct = 1;
		} else {
            if ((set = setmode(mode)) == NULL) {
				warn("Invalid file mode: %s", mode);
                fprintf(stderr, "\r");
                return 0;
            }
			oct = 0;
		}
#ifdef __APPLE__
	}
#endif /* __APPLE__*/
    if ((ftsp = fts_open(++argv, fts_options, 0)) == NULL) {
		warn("fts_open");
        fprintf(stderr, "\r");
        return 0;
    }
	for (rval = 0; (p = fts_read(ftsp)) != NULL;) {
		switch (p->fts_info) {
		case FTS_D:
			if (!Rflag)
				(void)fts_set(ftsp, p, FTS_SKIP);
			break;
		case FTS_DNR:			/* Warn, chmod, continue. */
			warnx("%s: %s", p->fts_path, strerror(p->fts_errno));
            fprintf(stderr, "\r");
			rval = 1;
			break;
		case FTS_DP:			/* Already changed at FTS_D. */
			continue;
		case FTS_NS:
			if (acloptflags & ACL_FLAG) /* don't need stat for -N */
				break;
		case FTS_ERR:			/* Warn, continue. */
			warnx("%s: %s", p->fts_path, strerror(p->fts_errno));
            fprintf(stderr, "\r");
			rval = 1;
			continue;
		case FTS_SL:			/* Ignore. */
		case FTS_SLNONE:
			/*
			 * The only symlinks that end up here are ones that
			 * don't point to anything and ones that we found
			 * doing a physical walk.
			 */
			if (!hflag)
				continue;
			/* else */
			/* FALLTHROUGH */
		default:
			break;
		}
#ifdef __APPLE__
/* If an ACL manipulation option was specified, manipulate */
		if (acloptflags & ACL_FLAG)	{
			if (0 != modify_file_acl(acloptflags, p->fts_accpath, acl_input, (int)aclpos, inheritance_level, !hflag))
				rval = 1;
		}
		else {
#endif /* __APPLE__ */
			newmode = oct ? omode : getmode(set, p->fts_statp->st_mode);
			if ((newmode & ALLPERMS) == (p->fts_statp->st_mode & ALLPERMS))
				continue;
			if ((*change_mode)(p->fts_accpath, newmode) && !fflag) {
				warn("Unable to change file mode on %s", p->fts_path);
                fprintf(stderr, "\r");
				rval = 1;
			} else {
				if (vflag) {
					(void)printf("%s", p->fts_accpath);

					if (vflag > 1) {
						char m1[12], m2[12];
						
						strmode(p->fts_statp->st_mode, m1);
						strmode((p->fts_statp->st_mode &
							 S_IFMT) | newmode, m2);
						
						(void)printf(": 0%o [%s] -> 0%o [%s]",
							     p->fts_statp->st_mode, m1,
					    (p->fts_statp->st_mode & S_IFMT) |
							     newmode, m2);
					}
					(void)printf("\n\r");
				}
				
			}
#ifdef __APPLE__
		}
#endif /* __APPLE__*/
	}
	if (errno)
		err(1, "fts_read");
#ifdef __APPLE__
	if (acl_input)
		acl_free(acl_input);
	if (mode && (acloptflags & ACL_FROM_STDIN))
		free(mode);
	
#endif /* __APPLE__ */
	if (set)
		free(set);
	return(rval);
}

void
chmod_usage(void)
{
#ifdef __APPLE__
	(void)fprintf(stderr,
		      "\rusage:\tchmod [-fhv] [-R [-H | -L | -P]] [-a | +a | =a  [i][# [ n]]] mode|entry file ...\n\r"
		      "\tchmod [-fhv] [-R [-H | -L | -P]] [-E | -C | -N | -i | -I] file ...\n\r"); /* add -A and -V when implemented */
#else
	(void)fprintf(stderr,
	    "\rusage: chmod [-fhv] [-R [-H | -L | -P]] mode file ...\n\r");
#endif /* __APPLE__ */
	// exit(1);
}
